app/src/main/java/de/vectordata/skynet/ui/chat/formatting/MarkdownFormatting.java

```diff
@@ -0,0 +1,107 @@
+package de.vectordata.skynet.ui.chat.formatting;
+
+import android.graphics.Typeface;
+import android.text.SpannableStringBuilder;
+import android.text.Spanned;
+import android.text.style.StrikethroughSpan;
+import android.text.style.StyleSpan;
+import android.text.style.TypefaceSpan;
+
+import java.util.HashMap;
+import java.util.Map;
+
+public class MarkdownFormatting implements IMessageFormatting {
+
+    @Override
+    public void format(SpannableStringBuilder builder) {
+        Map<Formatting, FormatState> formatStates = new HashMap<>();
+        for (Formatting formatting : Formatting.values())
+            formatStates.put(formatting, new FormatState(false, 0));
+
+        for (int i = 0; i < builder.length(); i++) {
+            char chr = builder.charAt(i);
+
+            Map.Entry<Formatting, FormatState> state = getState(formatStates, chr);
+            if (state == null) // Not a formatting-related char
+                continue;
+
+            if (state.getValue().isOpen()) {
+                // Close it
+                builder.setSpan(state.getKey().toSpanObject(), state.getValue().getStartIndex(), i, Spanned.SPAN_INCLUSIVE_INCLUSIVE);
+                state.getValue().end();
+            } else {
+                // Open it
+                state.getValue().begin(i);
+            }
+        }
+    }
+
+    private Map.Entry<Formatting, FormatState> getState(Map<Formatting, FormatState> states, char chr) {
+        for (Map.Entry<Formatting, FormatState> state : states.entrySet()) {
+            if (state.getKey().getFlagChar() == chr) {
+                return state;
+            }
+        }
+        return null;
+    }
+
+    private enum Formatting {
+        Bold('*'),
+        Italic('_'),
+        Strikeout('~'),
+        SourceCode('`');
```

Comment:
What do you think of implementing a subset of [CommonMark](https://commonmark.org/) instead of simplified markdown? In my opinion this would lead to less confusion and could be extended easily.

Reply:
I wanted to keep it simple because CommonMark parsers are usually not the lightest of libraries. Also, I don't think many people will need paragraph headers and tables in their messages. Maybe we can find a lightweight implementation of something similar in a library, but I haven't found one which is why i threw together this parser.

Reply:
I see. If we change to CommonMark in the future, the rendering of single asterisk will change from bold to italic but as chat message usually short-lived that won't be a big problem.